### PR TITLE
Extend ExecutionMode with DDL, Spark and Client only queries, dump ExecutionEngine

### PR DESCRIPTION
**Pull request by frsv**
Also remove IsMapReduce and IsTez, as it can be calculated from the ExecutionMode / ExecutionEngine.

### CODE REVIEW

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java [hunk 1]
```diff
@@ -154,29 +154,34 @@ private static Optional<String> dumpMapReduceCounters() {
...
+      List<? extends Iterable<G>> list,
+      Function<C, String> nameFn,
+      Function<C, Long> valueFn,
+      Function<G, String> displayNameFn) {
     JSONArray outerObj = new JSONArray();
 
-    list
-        .forEach(counters -> {
+    list.forEach(
+        counters -> {
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java:184]: Optional: I would suggest to move the formatting changes to a separate PR (possibly reformatting the whole codebase), especially since there are many such changes in this PR and they mix with the actual feature changes.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java [hunk 2]
```diff
@@ -239,20 +244,32 @@ private static String getRequestUser(HookContext hookContext) {
...
+  private static ExecutionMode getExecutionMode(QueryPlan plan) {
+    List<ExecDriver> mrTasks = Utilities.getMRTasks(plan.getRootTasks());
+    List<TezTask> tezTasks = Utilities.getTezTasks(plan.getRootTasks());
+    List<SparkTask> sparkTasks = Utilities.getSparkTasks(plan.getRootTasks());
+    List<DDLTask> ddlTasks = TasksRetriever.getDdlTasks(plan.getRootTasks());
+
+    // Utilities methods check for null, so possibly it is nullable
+    if (plan.getRootTasks() != null && plan.getRootTasks().isEmpty()) {
+      return ExecutionMode.CLIENT_ONLY;
+    } else if (!tezTasks.isEmpty()) {
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java:256]: `else` is not necessary if the body of the `if` above ends with `return`. Similarly for the other elses below.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java [hunk 3]
```diff
@@ -239,20 +244,32 @@ private static String getRequestUser(HookContext hookContext) {
     return requestUser == null ? hookContext.getUgi().getUserName() : requestUser;
   }
 
-  private static ExecutionMode getExecutionMode(List<ExecDriver> mrTasks, List<TezTask> tezTasks) {
-    if (tezTasks.size() > 0) {
+  private static ExecutionMode getExecutionMode(QueryPlan plan) {
+    List<ExecDriver> mrTasks = Utilities.getMRTasks(plan.getRootTasks());
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java:248]: Optional: If `tezTasks` is not empty, it's not necessary to compute `mrTasks`. I would suggest to move `mrTasks` variable definition below the `if (!tezTasks.isEmpty()) { ... }` instruction.

Similarly for `sparkTasks` and `ddlTasks`.

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/utils/TasksRetriever.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+public final class TasksRetriever {
+  private TasksRetriever() {}
+
+  public static List<DDLTask> getDdlTasks(List<Task<? extends Serializable>> tasks) {
+    List<DDLTask> ddlTasks = new ArrayList<>();
+    if (tasks != null) {
+      getDdlTasksRecursively(tasks, ddlTasks);
+    }
+
+    return ddlTasks;
```
Comment (dawidxc) [inline src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/utils/TasksRetriever.java:35]: The `ddlTasks` list will be iterated over multiple times, if there are dependent tasks for multiple tasks. What is the estimated max size of this list and the estimated number of tasks with dependent tasks?

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java [hunk 5]
```diff
@@ -229,20 +260,26 @@ private TezTask createTezTaskWithCounters(String id, CountersHolder counters) {
...
+        .groups()
+        .forEach(
+            group -> {
+              Group countersGroup = counters.getGroup(group.name());
+              group
+                  .counters()
+                  .forEach(
+                      (key, value) -> {
+                        Counter counterForName = countersGroup.getCounterForName(key);
+                        counterForName.setValue(value);
```
Comment (dawidxc) [inline src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java:273]: `counterForName` variable is redundant, I would suggest to inline it.

**Review by dawidxc — APPROVED** [review]

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java (hunk 1 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java:184]: Created #29

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/utils/TasksRetriever.java (hunk 4 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/utils/TasksRetriever.java]: I initially copied the method body from the `Utilities.getMrTasks`, but since we don't really need to have an actual list of tasks here, I changed it to `hasDdlTask`.

Answering the exact question – based on my experiments, DDL queries always have 1 DDLTask without dependencies, but I wouldn't rely on that.

File: src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java (hunk 5 above)
Comment (frsv) [reply src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructorTest.java]: Done

File: src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java (hunk 3 above)
Comment (frsv) [reply src/java/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventRecordConstructor.java]: Done
